### ISSUE TITLE
Implement AI/ML Module for AgriSave Backend

### ISSUE THREAD

**Issue by Fidelisaboke** [Fidelisaboke/AgriSave#4]
## Description

Develop and integrate the AI/ML module within the Django backend to provide intelligent predictions and recommendations for farmers.
The module should handle model training, loading, and inference through Django REST API endpoints.

This issue covers:

1. Setting up the ML module (`ml_engine/`) in Django.
2. Preparing and cleaning sample datasets.
3. Implementing baseline ML models.
4. Creating REST API endpoints for inference.
5. Integrating responses with the main API layer.

## Type of Change
- [x] New Feature

### **1. Setup**

* [x] Create `ml_engine/` Django app.
* [x] Configure folder structure:

  ```
  ml_engine/
  ├── __init__.py
  ├── models/                 # ML models (not Django ORM)
  ├── services/               # Inference logic
  ├── utils/                  # Preprocessing helpers
  ├── views.py
  ├── urls.py
  └── tests/
  ```
* [x] Install required packages:

  ```bash
  pip install scikit-learn tensorflow numpy pandas pillow joblib
  ```

---

### **2. Datasets & Preprocessing**

* [x] Add small sample datasets (subset of PlantVillage, NASA POWER climate data, FAO soil data).
* [x] Implement dataset loaders and preprocessing scripts in `ml_engine/utils/preprocess.py`.
* [x] Document dataset schema and features in `docs/data_notes.md`.

---

### **3. Model Implementation**

* [x] Implement **Disease Detection Model** (lightweight CNN or MobileNetV2 fine-tuned).
* [x] Implement **Climate Predictor** (LSTM or regression for temperature/rainfall trends).
* [x] Implement **Crop Recommender** (Random Forest based on soil + weather data).
* [x] Save models with `joblib` or TensorFlow `.h5` format.
* [x] Include model metadata (version, accuracy, last trained).

---

### **4. Inference Endpoints**

* [x] Create endpoints in `ml_engine/views.py`:
  * `/api/ml/predict-disease/`
  * `/api/ml/recommend-crop/`
  * `/api/ml/climate-forecast/`
* [x] Integrate with DRF views or routers.
* [x] Add mock responses first, then connect real predictions.
* [x] Validate inputs and handle exceptions gracefully.

---

### **5. Testing & Validation**

* [ ] Write unit tests for model loading, preprocessing, and prediction.
* [x] Test endpoints using Postman.
* [ ] Verify model response speed and JSON schema.
* [ ] Optimize inference for low latency.

---

### **6. Documentation**

* [x] Create `docs/ml_module.md` describing model architecture, training process, and usage.
* [x] Update API documentation to include new `/api/ml/` routes.

##  **Deliverables**

* [x] Functional ML models (CNN, RF, LSTM/regression).
* [x] Working Django API endpoints for predictions.
* [x] Documented dataset and inference flow.
* [ ] Test coverage for ML-related code.

 ### ⏰ **Estimated Time**: 
* **2–3 days** (iterative refinement on Wednesday–Thursday).
